nymea/notifications.py

```diff
@@ -18,53 +18,185 @@
 #  You should have received a copy of the GNU General Public License      #
 #  along with nymea-cli. If not, see <http://www.gnu.org/licenses/>.      #
 #                                                                         #
-# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # 
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
 
 import sys
 import socket
 import json
-import select
-import telnetlib
+import curses
+import os
+import ssl
 
 import nymea
 
-def notification_sniffer(nymeaHost, nymeaPort):
-    global commandId
-    
-    commandId = 0
-    print "Connecting notification handler..."
+def notification_sniffer(host, port):
+    global notificationSocket
+    global token
+    global initialSetupRequired
+    global pushButtonAuthAvailable
+    global authenticationRequired
+
     try:
-        tn = telnetlib.Telnet(nymeaHost, nymeaPort)
-    except :
-        print "ERROR: notification socket could not connect the to nymea-server. \n"
+        notificationSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        notificationSocket.connect((host, port));
+
+        # Perform initial handshake
+        try:
+            handshakeMessage = send_command("JSONRPC.Hello", {})
+        except:
+            context = ssl.create_default_context()
+            context.check_hostname = False
+            context.verify_mode = ssl.CERT_NONE
+            try:
+                notificationSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                notificationSocket.connect((host, port));
+                notificationSocket = context.wrap_socket(sock)
+                handshakeMessage = send_command("JSONRPC.Hello", {})
+            except:
+                print("SSL handshake failed.")
+
+        handshakeData = handshakeMessage['params']
+        print_json_format(handshakeData)
+        print("Connected to", handshakeData["server"], "\nserver version:", handshakeData["version"], "\nprotocol version:", handshakeData["protocol version"], "\n")
+
+        initialSetupRequired  = handshakeData['initialSetupRequired']
+        authenticationRequired = handshakeData['authenticationRequired']
+        pushButtonAuthAvailable = handshakeData['pushButtonAuthAvailable']
+
+        data = ''
+        while "}\n" not in data:
+            chunk = notificationSocket.recv(4096)
+            if chunk == '':
+                raise RuntimeError("socket connection broken")
+            data += chunk
+
+        dataJson = json.loads(data)
+        data = ''
+        if 'notification' in dataJsondataJson:
+            print(dataJson)
+
+        # If we don't need any authentication, we are done
```

Comment:
hmm... so the cli stores a token too, does it? I don't think the notification sniffer needs to handle authentication so I suppose most of this commented code can just go away, but looks like it doesn't work yet with out, am I reading that correct?

Reply:
unfortunatly, nymea-cli stores nothing. this is still not working as expected and would require quiet a rework to make it properly working. One centra socket for the entire application would make live much easier within the entire application...it has grown this direction over the years...

Reply:
ack.